source/postprocess/particle_density_histogram.cc

```diff
@@ -0,0 +1,268 @@
+                }                    
+              }    
+              /*
+              take the ratio between the actual error and the worst case 
+              error, resulting in a score from 0 to 1 for the cell
+              */
+              double distribution_score_current_cell = actual_error_squared/worst_case_error_squared;
+              //average
+              global_score += distribution_score_current_cell;
+              //max
+              if (distribution_score_current_cell >local_max_score){
+                local_max_score = distribution_score_current_cell;
+              }
+              //min
+              if (distribution_score_current_cell < local_min_score){
+                local_min_score = distribution_score_current_cell;
+              }
+            }
+          }
+      }
+
+      //get final values from all processors
+      double global_max_score = Utilities::MPI::max (local_max_score, this->get_mpi_communicator());
+      double global_min_score = Utilities::MPI::min (local_min_score, this->get_mpi_communicator());
+      double summed_score = Utilities::MPI::sum (global_score, this->get_mpi_communicator());
+      double global_cells_with_particles = Utilities::MPI::sum (cells_with_particles, this->get_mpi_communicator());
+      double average_score = summed_score / global_cells_with_particles;
+
+      // write to statistics file
+      statistics.add_value ("Minimal density gradient score: ", global_min_score);
+      statistics.add_value ("Average density gradient score: ", average_score);
+      statistics.add_value ("Maximal density gradient score: ", global_max_score);
+
+      std::ostringstream output;
+      output << global_min_score << "/ " <<average_score << "/ " << global_max_score;
+
+      return std::pair<std::string, std::string> ("Particle density gradient score min/avg/max:",
+                                                  output.str());
+    }
+
+
+
+    template <int dim>
+    std::list<std::string>
+    ParticleDensityStatistics<dim>::required_other_postprocessors() const
+    {
+      return {"particles"};
+    }
+
+    
+    template <int dim>
+    void ParticleDensityStatistics<dim>::sortParticles(const typename Triangulation<dim>::active_cell_iterator &cell,Table<dim,unsigned int> &buckets,double bucket_width)
+    {
+      for (unsigned int particle_manager_index = 0; particle_manager_index < this->n_particle_managers(); ++particle_manager_index)
+      {
+        //sort only the particles within the cell 
+        auto first_particle_in_cell = this->get_particle_manager(particle_manager_index).get_particle_handler().particles_in_cell(cell).begin();
+        auto last_particle_in_cell = this->get_particle_manager(particle_manager_index).get_particle_handler().particles_in_cell(cell).end();
+        
+        for(auto particle_iterator=first_particle_in_cell; particle_iterator != last_particle_in_cell; std::advance(particle_iterator,1))
```

Comment:
this is not wrong, but it can be simplified to:
```suggestion
        // sort the particles within the current cell 
        const typename ParticleHandler<dim>::particle_iterator_range particle_range =
        this->get_particle_manager(particle_manager_index).get_particle_handler().particles_in_cell(cell);

            for (const auto &particle: particle_range)
            {
            const double particle_x = particle.get_reference_location()[0];
            ...
            
```